### PR TITLE
CMake modernization

### PR DESCRIPTION
**Pull request by qiao-bo**
Refactor and update CMake files:

- replace the usage of directory-oriented old commands with target-oriented new commands.
   for example: use `target_include_directories()` instead of `include_directories()`, use `target_xxx()` wherever possible.
- minimize the usage of global variables.
- model target dependencies explicit with `target_link_libraries()`. 
   for example in libs, header include is defined per target, instead of everyone seeing everything. 
- add scope for target properties, separate PRIVATE,PUBLIC
- remove some old tweaks that are (probably) no longer needed.

TODO:

- [x] top level CMakeLists.txt
- [x] /compiler, /tools, and /libs subdirectories CMakeLists.txt
- [x] /cmake/Hipacc-functions.cmake
- [x] /samples-public/CMakeLists.txt

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,122 +1,45 @@
...
+        HOMEPAGE_URL ${HIPACC_GIT_REPOSITORY}
+        LANGUAGES CXX)
 
+# global setup for hipacc
 if(MSVC)
-    add_compile_options("/W3")
-    # move warnings from LLVM 8.0.1 to warning level 4
-    add_compile_options("/w44141" "/w44146" "/w44244" "/w44267" "/w44291")
+  add_compile_options("/W3")
+  # move warnings from LLVM 8.0.1 to warning level 4
```
Comment (oreiche) [inline CMakeLists.txt:23]: We are using LLVM 10 now do we? Maybe remove the version completely.

Comment (oreiche) [inline CMakeLists.txt]: We are using LLVM 10 now do we? Maybe remove the version completely.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,122 +1,45 @@
...
+  add_compile_options("/w44141" "/w44146" "/w44244" "/w44267" "/w44291")
```
Comment (oreiche) [inline CMakeLists.txt]: I know, it originally was my code, but the quotes are not necessary here.

File: CMakeLists.txt [hunk 3]
```diff
@@ -164,47 +80,28 @@ elseif(MSVC)
...
+#      set(${variable} "${${variable}} ${value}" PARENT_SCOPE)
+#    endforeach(variable)
+#  endif()
+#endfunction()
+#
+#if(NOT WIN32 AND NOT CYGWIN)
+#    # MinGW warns if -fvisibility-inlines-hidden is used.
+#    check_cxx_compiler_flag("-fvisibility-inlines-hidden" SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG)
+#    append_if(SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG "-fvisibility-inlines-hidden" CMAKE_CXX_FLAGS)
+#endif()
```
Comment (oreiche) [inline CMakeLists.txt]: I'm pretty sure that Cygwin is not a target that we still support. (this was from back in the days when it was a pain to compile "modern" C++11 code with MSVC)

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,122 +1,45 @@
...
-
-message("This is Hipacc ${HIPACC_VERSION}")
-
-#set(CMAKE_VERBOSE_MAKEFILE ON)
-set(CMAKE_SKIP_BUILD_RPATH FALSE)
-set(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
-set(CMAKE_INSTALL_RPATH ${CMAKE_INSTALL_PREFIX}/bin)
-set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
-set(CMAKE_INSTALL_MESSAGE LAZY)
-set(CMAKE_CXX_STANDARD 14)
```
Comment (oreiche) [inline CMakeLists.txt:60]: Question: Shouldn't the `CMAKE_CXX_STANDARD` stay? After all, this is a setting that affects the entire project.

File: CMakeLists.txt [hunk 5]
```diff
@@ -164,47 +80,28 @@ elseif(MSVC)
...
+#    # MinGW warns if -fvisibility-inlines-hidden is used.
+#    check_cxx_compiler_flag("-fvisibility-inlines-hidden" SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG)
+#    append_if(SUPPORTS_FVISIBILITY_INLINES_HIDDEN_FLAG "-fvisibility-inlines-hidden" CMAKE_CXX_FLAGS)
+#endif()
 
 # enable static code analysis by cppcheck
 include(CppcheckTargets OPTIONAL RESULT_VARIABLE CppcheckTargets_FOUND)
 if(CppcheckTargets_FOUND)
-    LIST(APPEND CPPCHECK_CUSTOM_ARGUMENTS --std=c++11 -DCLANG_VERSION_MAJOR=10 -D__clang__ -DCUDA_VERSION=10000)
+    LIST(APPEND CPPCHECK_CUSTOM_ARGUMENTS --std=c++14 -DCLANG_VERSION_MAJOR=10 -D__clang__ -DCUDA_VERSION=10000)
```
Comment (oreiche) [inline CMakeLists.txt]: If `CMAKE_CXX_STANDARD` was global, we could do `--std=c++${CMAKE_CXX_STANDARD}` here.

File: include/hipacc/AST/ASTFuse.h [hunk 6]
```diff
@@ -41,14 +41,14 @@
...
-#include "hipacc/DSL/ClassRepresentation.h"
-#include "hipacc/Vectorization/SIMDTypes.h"
+#include "KernelStatistics.h"
+#include "HostDataDeps.h"
+#include "ASTNode.h"
+#include "ASTTranslate.h"
+#include "CompilerOptions.h"
+#include "Builtins.h"
+#include "ClassRepresentation.h"
+#include "SIMDTypes.h"
```
Comment (oreiche) [inline include/hipacc/AST/ASTFuse.h]: Question: Why go for the local-include-approach? To be honest, I'm more in favor of the absolute path approach, going even further than before and including by using the full path from the workspace root e.g., `include/hipacc/AST/ASTNode.h`. This way its more or less guaranteed to have a unique path for each header.

File: lib/Backend/CMakeLists.txt [hunk 7]
```diff
@@ -1,44 +1,44 @@
...
+  PRIVATE VectorizationAST.cpp
+          Vectorizer.cpp
+          CPU_x86.cpp
+          CUDA.cpp
+          OpenCL_ACC.cpp
+          OpenCL_CPU.cpp
+          OpenCL_GPU.cpp
+          ClangASTHelper.cpp
+          InstructionSets.cpp)
+target_include_directories(hipaccBackend SYSTEM
```
Comment (oreiche) [inline lib/Backend/CMakeLists.txt:18]: The Hipacc includes are not system include paths. It guess we should remove the `SYSTEM` keyword here?

File: lib/Runtime/CMakeLists.txt [hunk 8]
```diff
@@ -1,24 +1,35 @@
...
+
+# this library is not installed because the static runtime lib has been skipped
+# in favour of the header-only runtime. this library is only build to detect
+# compilation issues during the build of Hipacc
+
+add_library(hipaccRuntime Base.cpp CPU.cpp)
+add_cppcheck(hipaccRuntime POSSIBLE_ERROR FAIL_ON_WARNINGS)
+target_include_directories(hipaccRuntime SYSTEM
+  PRIVATE ${LLVM_INCLUDE_DIRS}
+          $<$<BOOL:${CUDA_FOUND}>:${CUDA_INCLUDE_DIRS}>
```
Comment (oreiche) [inline lib/Runtime/CMakeLists.txt]: Nice!

File: compiler/CMakeLists.txt [hunk 9]
```diff
@@ -1,40 +1,28 @@
...
+  PRIVATE ${LLVM_DEFINITIONS})
+target_link_libraries(hipacc
+  PRIVATE hipaccRewrite
+          hipaccBackend
+          hipaccCompilerOptions
+  PRIVATE clangFrontend
+          clangDriver
+          clangLex)
+target_compile_features(hipacc
+  PRIVATE cxx_std_14)
```
Comment (oreiche) [inline compiler/CMakeLists.txt]: If `CMAKE_CXX_STANDARD` were global, we could omit this.

**Review by oreiche — COMMENTED** [review]
Apart from some suggestions, it looks good to me. This was long overdue, thanks for putting the effort in modernizing this!

File: CMakeLists.txt (hunk 3 above)
Comment (qiao-bo) [reply CMakeLists.txt]: ok, thanks, good to know

File: CMakeLists.txt (hunk 5 above)
Comment (qiao-bo) [reply CMakeLists.txt]: this was a question that i ask myself, do you think there will be future scenarios that targets using different language std? like some c++11, another c++17?

Comment (richardmembarth) [reply CMakeLists.txt]: I think we should have a minimum standard `CMAKE_CXX_STANDARD` hat is set for the project (compiler). The runtime might require a different standard.

File: include/hipacc/AST/ASTFuse.h (hunk 6 above)
Comment (qiao-bo) [reply include/hipacc/AST/ASTFuse.h]: correct, but the old approach has one disadvantage: if i mistakenly introduce some circular dependencies on the header from two targets. CMake during generation phase will not be able to detect this. If we keep include local to targets, the dependency is modeled explicitly. Then during Cmake generation this can already be detected.
This is the main reason i switched to local-include-approach. Correct me if i am wrong

File: CMakeLists.txt (hunk 5 above)
Comment (qiao-bo) [reply CMakeLists.txt]: what if later someone wants to link some external library?  such as boost with c++17

File: lib/Backend/CMakeLists.txt (hunk 7 above)
Comment (qiao-bo) [reply lib/Backend/CMakeLists.txt:18]: ${LLVM_INCLUDE_DIRS) is?

File: CMakeLists.txt (hunk 5 above)
Comment (qiao-bo) [reply CMakeLists.txt]: or vice versa, use an external library with older standard ;)

File: lib/Backend/CMakeLists.txt (hunk 7 above)
Comment (richardmembarth) [reply lib/Backend/CMakeLists.txt:18]: Yes and no ... basically, the rationale behind using `SYSTEM` was to suppress warnings from LLVM includes.
`SYSTEM` is ignored on Windows and might cause problems on Linux depending on the LLVM installation directory.

File: include/hipacc/AST/ASTFuse.h (hunk 6 above)
Comment (richardmembarth) [reply include/hipacc/AST/ASTFuse.h]: @oreiche you mean to include `hipacc/AST/ASTNode.h`?
This makes sense.

File: CMakeLists.txt (hunk 5 above)
Comment (richardmembarth) [reply CMakeLists.txt]: The standard is set for Cppcheck here, so no problem.
For older standards, there exist special flags to specify compatibility, like `_GLIBCXX_USE_CXX11_ABI`.

Comment (qiao-bo) [reply CMakeLists.txt]: @richardmembarth, ah ok, i thought we were taking about a global ${CMAKE_CXX_STANDARD} for the whole project.

Comment (oreiche) [reply CMakeLists.txt]: Richard is right. The entire project should be the same standard. The runtime, being a header-only library, is actually not affected by the project-wide standard and could be less restrictive (I guess it should compile fine with C++11 right?). So we have two standards, one for the project and one as minimum for the runtime.

I don't really recall which code parts are processed by this `cpp-check`. If the runtime is checked as well, we need to take care that the standard is used that we agreed on for the runtime.

Regarding the standard mismatch issue. I think that a dependency using a lower standard is not a problem. If the dependency is using a higher standard and exposes that via its public interface (e.g., `std::filesystem`) we would need to upgrade our project's standard (or write a interface wrapper), but I guess thats also fine.

Comment (qiao-bo) [reply CMakeLists.txt]: ok, i will revert to a global setting for the cxx standard

File: include/hipacc/AST/ASTFuse.h (hunk 6 above)
Comment (qiao-bo) [reply include/hipacc/AST/ASTFuse.h]: @richardmembarth @oreiche Let me rephrase it if it was not clear. let's say i have target A and B. 
- If i use global include (old approach), i can do:
   in targetA.h, #include"hipacc/targetB/targetB.h" and
   in targetB.h, #include"hipacc/targetA/targetA.h".

   Then, there is no way to build it. But "cmake .." will still run successfully. although build will fail later.

- if i use local-include-approach, 
   in targetA.h, To be able to #include"targetB.h", I need to target_link_library(targetA, private targetB), and
   in targetB.h, To be able to #include"targetA.h", I need to target_link_library(targetB, private targetA),

  Then "cmake .." will NOT run, because of something like "The inter-target dependency graph contains the following strongly connected component (cycle):". Then you will get a quite clear idea of which targets are the problem. 

I agree at this stage Hipacc does not necessary need it. It is not that complicated anyway. But it's about reducing the chance to introduce errors.

Comment (oreiche) [reply include/hipacc/AST/ASTFuse.h]: @qiao-bo Circular includes should be mitigated by include guards anyway. The problem you are describing sounds like its cause by a naming conflict?

@richardmembarth No, actually I really meant `include/hipacc/...`. Its a matter of taste, but I was recently told that the "Google way" is to include project-owned non-public headers by their "absolute path" in the project (meaning relative path from project's root). This way, you have a unique path for every include.

Example structure:
~~~
include/foo/bar.hpp
src/baz.hpp
src/baz.cpp
~~~

In `baz.cpp`:
~~~cpp
#include "foo/bar.hpp" // include public header, but could also be "include/foo/bar.hpp", as it is project owned
#include "src/baz.hpp" // include private header
~~~

And the include guards in the headers should encode the header path to some degree to be unique as well. (e.g., `SRC_BAZ_HPP` and `INCLUDE_FOO_BAR_HPP`)

Comment (richardmembarth) [reply include/hipacc/AST/ASTFuse.h]: I'm a fan of `foo/bar.hpp` instead of `include/foo/bar.hpp`.
LLVM handles this the same way.

Comment (oreiche) [reply include/hipacc/AST/ASTFuse.h]: Note that `ASTNode.h` is not a public header, so according to my comment above, I would always include it via its full path.

Comment (oreiche) [reply include/hipacc/AST/ASTFuse.h]: @richardmembarth `foo/bar.hpp` is also fine for me. Thats sufficiently unique enough in our context.

@qiao-bo I really do like the explicitness you try to introduce here. However, by this approach, you kind of mix linking libraries and including headers. In many cases, you will end up with circular header includes, that are prevented by include guards and might need some forward declarations. After all, its perfectly fine to have two libraries depending on each other include-wise, which would not be possible with your approach :/

Comment (qiao-bo) [reply include/hipacc/AST/ASTFuse.h]: ok, thanks for the discussion. I will change back to the "foo/bar.hpp" style, with include guards as well.

File: lib/Backend/CMakeLists.txt (hunk 7 above)
Comment (qiao-bo) [reply lib/Backend/CMakeLists.txt:18]: @oreiche @richardmembarth, do you suggest to use SYSTEM or not? I removed it and warnings from clang and llvm includes explode...

Comment (oreiche) [reply lib/Backend/CMakeLists.txt:18]: To be honest, I do not really understand the problem without knowing the exact warnings that are reported. If `SYSTEM` is the only way to get rid of them, then add it. However, adding them only for `${LLVM_INCLUDE_DIRS}` should be sufficient right?

Comment (richardmembarth) [reply lib/Backend/CMakeLists.txt:18]: The compiler is emitting warnings for the LLVM headers, not our code.
For this reason it makes sense to use `SYSTEM` for now.

Comment (qiao-bo) [reply lib/Backend/CMakeLists.txt:18]: yes, warnings are from LLVM and Clang headers, some examples:
```
/scratch-local/usr/local/include/llvm/Support/TrailingObjects.h: In instantiation of ‘static size_t llvm::TrailingObjects<BaseTy, TrailingTys>::callNum        TrailingObjects(const BaseTy*, llvm::trailing_objects_internal::TrailingObjectsBase::OverloadToken<BaseTy>) [with BaseTy = clang::ConstantArrayType; Tr        ailingTys = {const clang::Expr*}; size_t = long unsigned int]’:
/scratch-local/usr/local/include/llvm/Support/TrailingObjects.h:181:55:   required from ‘static NextTy* llvm::trailing_objects_internal::TrailingObject        sImpl<Align, BaseTy, TopTrailingObj, PrevTy, NextTy, MoreTys ...>::getTrailingObjectsImpl(BaseTy*, llvm::trailing_objects_internal::TrailingObjectsBase        ::OverloadToken<NextTy>) [with int Align = 8; BaseTy = clang::ConstantArrayType; TopTrailingObj = llvm::TrailingObjects<clang::ConstantArrayType, const         clang::Expr*>; PrevTy = clang::ConstantArrayType; NextTy = const clang::Expr*; MoreTys = {}]’ 

/scratch-local/usr/local/include/clang/AST/Expr.h: In constructor ‘clang::ImplicitCastExpr::ImplicitCastExpr(clang::ImplicitCastExpr::OnStack_t, clang:        :QualType, clang::CastKind, clang::Expr*, clang::ExprValueKind)’:
/scratch-local/usr/local/include/clang/AST/Expr.h:3288:30: warning: unused parameter ‘_’ [-Wunused-parameter] ImplicitCastExpr(OnStack_t _, QualType ty, CastKind kind, Expr *op
/scratch-local/usr/local/include/clang/AST/Expr.h: In constructor ‘clang::BinaryOperator::BinaryOperator(clang::Expr*, clang::Expr*, clang::BinaryOpera        tor::Opcode, clang::QualType, clang::ExprValueKind, clang::ExprObjectKind, clang::SourceLocation, clang::FPOptions, bool)’:
``` 
then i will use `SYSTEM` for now, only for `${LLVM_INCLUDE_DIRS)`.

Comment (oreiche) [reply lib/Backend/CMakeLists.txt:18]: Oh boy, what are they doing...? Doesn't clang use clang-tidy?? ^^

**Comment by qiao-bo** [comment]
@oreiche @richardmembarth I incorporated all the suggestions so far. thanks for the comments.
